feat: integrate creator pattern implementations in DeployImplementations.s.sol

stevennevins opened the pull request:
**Description**

Deploys the v2 implementations for FaultDisputeGameV2 and PermissionedDisputeGameV2 via DeployImplementations.s.sol

**Tests**

Added tests to DeployImplementations.t.sol

**Additional context**

N/A

**Metadata**

Closes [#17256](https://github.com/ethereum-optimism/optimism/issues/17256)

stevennevins committed:
feat: integration v2 implementations in DeployImplementations.s.sol + tests

stevennevins committed:
feat: add feature flag on and integrate deploy.s.sol

stevennevins committed:
feat: integrate deploy config and pass input struct values from config

stevennevins committed:
fix: configure a proposer in PDG

stevennevins committed:
fix: update the input struct for DeployImplementations.s.sol

stevennevins committed:
fix: implementation structs

mbaxter commented on op-chain-ops/interopgen/deploy.go:
```diff
@@ -209,7 +217,32 @@ func DeploySuperchainToL1(l1Host *script.Host, opcmScripts *opcm.Scripts, superC
 	// This could all be automatic once we have better output-contract typing/scripting
 	return &SuperchainDeployment{
-		Implementations:       Implementations(implementationsDeployment),
+		Implementations: Implementations{
```
Why doesn't the previous casting approach work here anymore?

stevennevins replied:
investigating. 

i think it might have been related to the structs being out of sync, but will double check

stevennevins replied:
Previous approach was actually fine still , nice! 
[2073042](https://github.com/ethereum-optimism/optimism/pull/17329/commits/207304217885f34b84b7c6a0c58a2aaf8a71049a)

mbaxter commented on op-deployer/pkg/deployer/bootstrap/implementations.go:
```diff
@@ -205,11 +205,19 @@ func Implementations(ctx context.Context, cfg ImplementationsConfig) (opcm.Deplo
-			SuperchainProxyAdmin:            cfg.SuperchainProxyAdmin,
-			UpgradeController:               cfg.UpgradeController,
-			Challenger:                      cfg.Challenger,
+			DeployV2DisputeGames:            false, // Default to false for backward compatibility
```
Do we need this new `DeployV2DisputeGames ` flag or can we use `DevFeatureBitmap`?

stevennevins replied:
we can use the bitmap, this was written before the bitmap feature was landed

stevennevins replied:
updated to use the bitmap now

mbaxter commented on op-deployer/pkg/deployer/opcm/implementations.go:
```diff
@@ -15,11 +15,19 @@ type DeployImplementationsInput struct {
-	SuperchainProxyAdmin            common.Address
-	UpgradeController               common.Address
-	Challenger                      common.Address
+	DeployV2DisputeGames            bool
```
Same question on whether we should use `DevFeatureBitmap` here instead.

stevennevins replied:
same as before, we can move to the bitmap

stevennevins replied:
updated to use the bitmap

mbaxter commented on op-deployer/pkg/deployer/opcm/implementations.go:
```diff
@@ -15,11 +15,19 @@ type DeployImplementationsInput struct {
+	SuperchainProxyAdmin  common.Address
+	UpgradeController     common.Address
+	Proposer              common.Address
+	Challenger            common.Address
```
Wondering why we didn't have a field for the `Proposer` previously.  It looks like this `Challenger` parameter was being injected into OPCMStandardValidator and that's all it was used for - is that right?  We should probably update the StandardValidator to check both challenger and proposer (could be a separate PR).

stevennevins replied:
yeah not clear to me why we didn't have a proposer field before

mbaxter commented on op-deployer/pkg/deployer/opcm/implementations_test.go:
```diff
@@ -66,11 +66,19 @@ func TestNewDeployImplementationsScript(t *testing.T) {
+			ProtocolVersionsProxy: protocolVersionsAddress,
+			SuperchainProxyAdmin:  proxyAdminAddress,
+			UpgradeController:     common.BigToAddress(big.NewInt(13)),
+			Proposer:              common.BigToAddress(big.NewInt(15)), // Added Proposer field
```
Extraneous comment :D
```suggestion
			Proposer:              common.BigToAddress(big.NewInt(15)),
```

stevennevins committed:
Update implementations_test.go

Co-authored-by: mbaxter <meredith@oplabs.co>

stevennevins committed:
fix: use previous direct casting appraoch

stevennevins committed:
feat: integrate new bitmap feature

stevennevins committed:
fix: delete old bool

stevennevins committed:
fix: lint

stevennevins committed:
feat: add proposer to the config

stevennevins committed:
fix: implementation struct usage in sepolia test

stevennevins committed:
fix: add proposer to the struct for integration tests

stevennevins committed:
fix: fund the proposer and add it to the supercahin config

stevennevins committed:
fix: constructor args for pdgV2

mbaxter commented on packages/contracts-bedrock/scripts/deploy/DeployConfig.s.sol:
```diff
@@ -76,6 +76,12 @@ contract DeployConfig is Script {
     uint256 public daResolverRefundPercentage;
 
+    // V2 Dispute Game Configuration
+    uint256 public faultGameV2MaxGameDepth;
+    uint256 public faultGameV2SplitDepth;
+    uint256 public faultGameV2ClockExtension;
+    uint256 public faultGameV2MaxClockDuration;
```
```suggestion
    uint256 public faultGameMaxGameDepth;
    uint256 public faultGameSplitDepth;
    uint256 public faultGameClockExtension;
    uint256 public faultGameMaxClockDuration;
```

stevennevins replied:
I was thinking we would align this when we remove the v2 contracts and replace the v1, wdyt?

Inphi replied:
I'm fine with keeping the V2 and removing it later so that it's unambiguous what these inputs are for. To avoid accidental misuse of V2 inputs for V1.

mbaxter replied:
All of this config is new since we were not doing any configuration when deploying the blueprints previously. Just seems like extra work to me to have to remove all instances of "v2"  and I don't see how it would be misused. 🤷‍♀️

stevennevins replied:
I think it's more about being able to search the project for disputeGame and it taking us to all the old instances of functions and config

stevennevins replied:
on second thought though it's not as easy to select the absence of the v2

mbaxter commented on packages/contracts-bedrock/scripts/deploy/DeployImplementations.s.sol:
```diff
@@ -464,6 +480,44 @@ contract DeployImplementations is Script {
         _output.anchorStateRegistryImpl = impl;
     }
 
+    function deployFaultDisputeGameV2Impl(Input memory _input, Output memory _output) private {
```
```suggestion
    function deployFaultDisputeGameImpl(Input memory _input, Output memory _output) private {
```

mbaxter commented on packages/contracts-bedrock/scripts/deploy/DeployImplementations.s.sol:
```diff
@@ -464,6 +480,44 @@ contract DeployImplementations is Script {
+        _output.faultDisputeGameV2Impl = impl;
+    }
+
+    function deployPermissionedDisputeGameV2Impl(Input memory _input, Output memory _output) private {
```
```suggestion
    function deployPermissionedDisputeGameImpl(Input memory _input, Output memory _output) private {
```

stevennevins committed:
revert: remove proposer changes from op-chain-ops and op-deployer

Reverted changes to op-chain-ops/interopgen and op-deployer/pkg/deployer
directories back to the develop branch version as these changes are no
longer necessary for this PR.

stevennevins committed:
fix: remove proposer from deploy and deploy implementations

mbaxter commented on packages/contracts-bedrock/test/opcm/DeployImplementations.t.sol:
```diff
@@ -117,6 +126,10 @@ contract DeployImplementations_Test is Test {
+            73, // faultGameV2MaxGameDepth
+            30, // faultGameV2SplitDepth
+            10800, // faultGameV2ClockExtension
+            302400, // faultGameV2MaxClockDuration
```
Looks like the intention of this test is to fuzz all of the configurable inputs so we should probably fuzz these even though figuring out the constraints will be a bit tricky.  And as part of that we should fuzz the devFeatureBitmap as well.

stevennevins replied:
yeah we would need to fuzz with the devFeatureBitmap because the way it is we wouldn't be deploying the v2 contracts anyway

stevennevins replied:
[b56a5d0](https://github.com/ethereum-optimism/optimism/pull/17329/commits/b56a5d0728edb435a6705fdb8c32f08126e3f3ec)

stevennevins committed:
fix: add back struct params

mbaxter commented on packages/contracts-bedrock/test/opcm/DeployImplementations.t.sol:
```diff
@@ -237,6 +256,79 @@ contract DeployImplementations_Test is Test {
         deployImplementations.run(input);
     }
 
+    function test_deployImplementation_withV2Enabled_succeeds() public {
```
There is this test above: `test_run_deployMipsV1OnMainnetOrSepolia_reverts` that validates we revert if bad config is used on sepolia or mainnet.  I think it probably makes sense to enforce that we can't deploy the v2 contracts to production for now, while development is ongoing.  We can remove this later.

stevennevins replied:
We have assertions for this deploying v1 vs v2 based on the feature flag bitmap and OPCM reverts if we deploy to a non anvil/testing environment with a non zero bitmap

if we want additional assurances we can probably duplicate the OPCM check where it reverts if the bitmap has any feature flags toggled on.  We should be covered from a safety perspective though with the OPCM check

mbaxter replied:
> OPCM reverts if we deploy to a non anvil/testing environment with a non zero bitmap

Where does that revert happen?

stevennevins replied:
https://github.com/ethereum-optimism/optimism/blob/8c3da2d814560450ba08fd08dae59a44a7678b34/packages/contracts-bedrock/src/L1/OPContractsManager.sol#L73

mbaxter replied:
Yeah - that check seems like it already covers this.

mbaxter commented on packages/contracts-bedrock/test/opcm/DeployImplementations.t.sol:
```diff
@@ -237,6 +256,79 @@ contract DeployImplementations_Test is Test {
+        assertNotEq(address(output.disputeGameFactoryImpl), address(0), "DisputeGameFactory should still be deployed");
+    }
+
+    function test_reuseImplementation_withV2Flags_succeeds() public {
```
I think its worth verifying that the various constructor parameters were set correctly on the deployed v2 contracts.

stevennevins replied:
on second thought, i think the intent is that we get those assurances from our other tests like the standard validator after we run the OPCM deployment

mbaxter replied:
I think the main use case for the StandardValidator is to provide a way to validate / check live OPCM contracts ([for example](https://github.com/ethereum-optimism/superchain-ops/blob/38b22d918d57b46b67f2214f37920f8aa92d368f/src/improvements/template/OPCMUpgradeV400.sol#L132-L134)).  But we should still be writing unit tests at a lower level.

stevennevins replied:
Yeah i think this is a bit of bigger scope than these tests we added, but if we are running it on a live environment it should be able to validate our test environment which is what a lot of these test are attempting even though some things are being mocked https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/test/L1/OPContractsManagerStandardValidator.t.sol

stevennevins replied:
Discussed offline to add the asserts until we have coverage later here

stevennevins replied:
[5b6ea66](https://github.com/ethereum-optimism/optimism/pull/17329/commits/5b6ea662b9fa2397961867d83fc720f7bc182eec)

mbaxter commented on packages/contracts-bedrock/scripts/deploy/DeployImplementations.s.sol:
```diff
@@ -79,6 +88,8 @@ contract DeployImplementations is Script {
         IAnchorStateRegistry anchorStateRegistryImpl;
         ISuperchainConfig superchainConfigImpl;
         IProtocolVersions protocolVersionsImpl;
+        IFaultDisputeGameV2 faultDisputeGameV2Impl;
+        IPermissionedDisputeGameV2 permissionedDisputeGameV2Impl;
```
I would avoid using "v2" unless we need it to distinguish from v1. Since these are new, unambiguous fields I think we can simplify to:
```suggestion
        IFaultDisputeGame faultDisputeGameV2Impl;
        IPermissionedDisputeGame permissionedDisputeGameV2Impl;
```
We'll eventually drop the "v2" so, this just reduces the clean-up we'll have to do later.

stevennevins committed:
fix: use casting for Impleemntations struct

mbaxter commented on op-deployer/pkg/deployer/bootstrap/implementations.go:
```diff
@@ -89,6 +93,19 @@ func (c *ImplementationsConfig) Check() error {
+	}
+	if c.FaultGameV2MaxClockDuration == 0 {
+		c.FaultGameV2MaxClockDuration = 302400
+	}
```
I don't see us setting defaults anywhere else in this method?  I think its unexpected that `Check` would modify the config - we should error here based on `DevFeatureBitmap`.

stevennevins replied:
Updated [b7b9320](https://github.com/ethereum-optimism/optimism/pull/17329/commits/b7b93206de6ad2a78ab5747df8de7a39dfa4b97b)

mbaxter commented on op-deployer/pkg/deployer/bootstrap/implementations.go:
```diff
@@ -39,6 +39,10 @@ type ImplementationsConfig struct {
+	FaultGameV2MaxGameDepth         uint64             `cli:"fault-game-v2-max-game-depth"`
+	FaultGameV2SplitDepth           uint64             `cli:"fault-game-v2-split-depth"`
+	FaultGameV2ClockExtension       uint64             `cli:"fault-game-v2-clock-extension"`
+	FaultGameV2MaxClockDuration     uint64             `cli:"fault-game-v2-max-clock-duration"`
```
```suggestion
	FaultGameMaxGameDepth         uint64             `cli:"fault-game-max-game-depth"`
	FaultGameSplitDepth           uint64             `cli:"fault-game-split-depth"`
	FaultGameClockExtension       uint64             `cli:"fault-game-clock-extension"`
	FaultGameMaxClockDuration     uint64             `cli:"fault-game-max-clock-duration"`
```

stevennevins committed:
Update packages/contracts-bedrock/scripts/deploy/DeployImplementations.s.sol

Co-authored-by: mbaxter <meredith@oplabs.co>

stevennevins committed:
Update op-deployer/pkg/deployer/bootstrap/implementations.go

Co-authored-by: mbaxter <meredith@oplabs.co>

Inphi commented on packages/contracts-bedrock/scripts/deploy/DeployImplementations.s.sol:
```diff
@@ -464,6 +479,44 @@ contract DeployImplementations is Script {
 
+    function deployFaultDisputeGameV2Impl(Input memory _input, Output memory _output) private {
+        IFaultDisputeGameV2.GameConstructorParams memory params;
+        params.gameType = GameType.wrap(0); // Default game type
```
```suggestion
        params.gameType = GameTypes.CANNON;
```

Inphi commented on packages/contracts-bedrock/scripts/deploy/DeployImplementations.s.sol:
```diff
@@ -464,6 +479,44 @@ contract DeployImplementations is Script {
+
+    function deployPermissionedDisputeGameV2Impl(Input memory _input, Output memory _output) private {
+        IFaultDisputeGameV2.GameConstructorParams memory params;
+        params.gameType = GameType.wrap(1); // Permissioned game type
```
```suggestion
        params.gameType = GameTypes.PERMISSIONED_CANNON;
```

Inphi commented on packages/contracts-bedrock/scripts/deploy/DeployImplementations.s.sol:
```diff
@@ -637,8 +690,37 @@ contract DeployImplementations is Script {
+                address(_output.permissionedDisputeGameV2Impl) != address(0),
+                "DeployImplementations: V2 flag enabled but PermissionedDisputeGameV2 not deployed"
+            );
+        } else {
```
```suggestion
```
The true branch of the if statement is already handled by the above call to `assertValidContractAddresses`.

stevennevins replied:
Updated 
[2459546](https://github.com/ethereum-optimism/optimism/pull/17329/commits/2459546df80f2fc2ed276b64b1af913b62f9ff50)

Inphi reviewed (commented):
Regarding naming. I think the V2 suffix should be used everywhere. So that once the old FDG is fully replaced, it's easy to know which area of the code needs to be cleaned up.

stevennevins committed:
fix: remove modification of inputs ad revert if feature flag enabled

stevennevins committed:
Update packages/contracts-bedrock/scripts/deploy/DeployImplementations.s.sol

Co-authored-by: Inphi <mlaw2501@gmail.com>

stevennevins committed:
Update packages/contracts-bedrock/scripts/deploy/DeployImplementations.s.sol

Co-authored-by: Inphi <mlaw2501@gmail.com>

stevennevins committed:
fix: go back to v2 impl

stevennevins committed:
fix: missing import and simplify conditional for valid impl addresses

stevennevins committed:
fix: update check for correct nibble after kona consumed 0x010

stevennevins committed:
fix: remove unused import

mbaxter commented:
> Regarding naming. I think the V2 suffix should be used everywhere. So that once the old FDG is fully replaced, it's easy to know which area of the code needs to be cleaned up.

I'm not sure this helps.  We're editing different areas of the code from what will need to be cleaned up.  For clean up, we'll have to remove some old (non-v2) params. I think the `DevFeatures` flag is what we'll need to track wrt clean-up.

Inphi commented:
> > Regarding naming. I think the V2 suffix should be used everywhere. So that once the old FDG is fully replaced, it's easy to know which area of the code needs to be cleaned up.
> 
> I'm not sure this helps. We're editing different areas of the code from what will need to be cleaned up. For clean up, we'll have to remove some old (non-v2) params. I think the `DevFeatures` flag is what we'll need to track wrt clean-up.

There are a couple areas where we introduce a new FDGV2 thing without a DevFeatures guard. There are a bunch of these in op-deployer. When we go and remove the non-v2 params, it'll be nice to easily know which of these should be cleaned up.

Inphi commented on op-deployer/pkg/deployer/bootstrap/implementations.go:
```diff
@@ -89,6 +93,25 @@ func (c *ImplementationsConfig) Check() error {
+	// DevFeatureBitmap 0x0100 bit enables V2 dispute game deployment
+	var deployV2FeatureFlag common.Hash
+	deployV2FeatureFlag[30] = 0x01
+	deployV2Games := (c.DevFeatureBitmap[30] & deployV2FeatureFlag[30]) != 0
```
This won't be the last time we'll be introducing dev feature flags. So lets introduce a constant for the deployV2Games feature flag as was done in solidity. And you could do a bitwise for a generic flag checker rather than check specific bits:
```go
func isDevFeatureEnabled(bitmap, flag common.Hash) bool {
	b := new(big.Int).SetBytes(bitmap[:])
	f := new(big.Int).SetBytes(flag[:])
	return new(big.Int).And(b, f).BitLen() != 0
}
```

stevennevins replied:
Do you mean define a constant in this file or elsewhere?

stevennevins replied:
put it here
[a9edb22](https://github.com/ethereum-optimism/optimism/pull/17329/commits/a9edb226fa2ebb1b737bcac2651779aa72f20d18)

stevennevins replied:
offline feedback

fixed in
[6856dff](https://github.com/ethereum-optimism/optimism/pull/17329/commits/6856dfff2215db81f3c150ce947fd213b08e519f)

Inphi reviewed (approved):
Looks good. Let's ensure that dev feature flag checks are generic in op-deployer.

stevennevins committed:
test: add assertions for constructor params

stevennevins committed:
test: add sensible fuzz input ranges

stevennevins committed:
feat: add reasonable constraints to deploys script and fuzz inputs

stevennevins committed:
fix: remove unused import

stevennevins committed:
refactor: add helper function and constant

stevennevins committed:
fix: move flags to their owner file

stevennevins committed:
fix: lint warning

mbaxter commented on packages/contracts-bedrock/scripts/deploy/DeployImplementations.s.sol:
```diff
@@ -104,6 +115,53 @@ contract DeployImplementations is Script {
         deployMipsSingleton(_input, output_);
         deployDisputeGameFactoryImpl(output_);
         deployAnchorStateRegistryImpl(_input, output_);
+        if (DevFeatures.isDevFeatureEnabled(_input.devFeatureBitmap, DevFeatures.DEPLOY_V2_DISPUTE_GAMES)) {
```
A couple questions:
- Not really opposed to doing extra validation here, but it doesn't follow the existing pattern.  It looks like we have input validation being handled in `op-deployer/pkg/deployer/bootstrap/implementations.go`.  Is the issue that those checks don't cover all invocations of this script?  If so, should all of those checks move here?
- Do we need really detailed validation here?  Seems like we should just be checking if values are set (non-zero) or not set (zero).

stevennevins replied:
yeah we don't really need these, we can just have non zero checks

stevennevins replied:
added back the checks, i noticed when going back through there was a check for exceeding the `challengePeriodSeconds` https://github.com/ethereum-optimism/optimism/blob/bd592e34554de1604ee063c57ead92512d364fbe/packages/contracts-bedrock/scripts/deploy/DeployImplementations.s.sol#L595-L597

mbaxter replied:
good catch! though I think that means these checks should move into `assertValidInput`

stevennevins replied:
sgtm, I was thinking that initially, and preferred having it all be in the conditional block. will refactor it to the other function

stevennevins replied:
Updated [0cd38ac](https://github.com/ethereum-optimism/optimism/pull/17329/commits/0cd38ac84359226f6d745992a787e092c3f461af)

mbaxter commented on packages/contracts-bedrock/test/opcm/DeployImplementations.t.sol:
```diff
@@ -78,36 +86,50 @@ contract DeployImplementations_Test is Test {
+                & (
+                    uint256(DevFeatures.OPTIMISM_PORTAL_INTEROP) | uint256(DevFeatures.CANNON_KONA)
+                        | uint256(DevFeatures.DEPLOY_V2_DISPUTE_GAMES)
                 )
```
(nit) I don't think we need to constrain the devFeatures bitmap to known flags - its probably a good idea to make sure nothing odd happens even if unused bits are set.

stevennevins replied:
Updated: [41c48e4](https://github.com/ethereum-optimism/optimism/pull/17329/commits/41c48e4509c126ed05272e1b613a38be8214736d)

mbaxter commented on packages/contracts-bedrock/test/opcm/DeployImplementations.t.sol:
```diff
@@ -78,36 +86,50 @@ contract DeployImplementations_Test is Test {
-        vm.assume(_proofMaturityDelaySeconds != 0);
-        vm.assume(_disputeGameFinalityDelaySeconds != 0);
+        _proofMaturityDelaySeconds = bound(_proofMaturityDelaySeconds, 1, 365 days);
+        _disputeGameFinalityDelaySeconds = bound(_disputeGameFinalityDelaySeconds, 1, 365 days);
```
(nit) I would probably avoid bounding parameters unless we have to in order to avoid reverts. I would think we want to cast a wide net here.

stevennevins replied:
Expanded, mainly was just bounding to reasonable values [3d51968](https://github.com/ethereum-optimism/optimism/pull/17329/commits/3d519688618815741bf0d05313664efe208d9ca6)

mbaxter commented on packages/contracts-bedrock/test/opcm/DeployImplementations.t.sol:
```diff
@@ -237,6 +313,198 @@ contract DeployImplementations_Test is Test {
+    function test_v2ParamsValidation_withHugeValues_reverts() public {
+        // When V2 flag is disabled, huge V2 params should be rejected
+        DeployImplementations.Input memory input = defaultInput();
+        input.devFeatureBitmap = bytes32(0); // V2 disabled
```
Seems more relevant to run these test with v2 enabled (or both enabled and disabled).

stevennevins replied:
Agree, updated and removed the checks when the feature flag isn't on since that scneario doesn't matter [150ccb5](https://github.com/ethereum-optimism/optimism/pull/17329/commits/150ccb5c727573b2c84365faebfa45fe3052450d)

mbaxter commented on packages/contracts-bedrock/test/opcm/DeployImplementations.t.sol:
```diff
@@ -237,6 +313,198 @@ contract DeployImplementations_Test is Test {
+        vm.expectRevert("DeployImplementations: V2 disabled but faultGameV2MaxClockDuration too large");
+        deployImplementations.run(input);
+
+        // Reset and test huge max game depth
```
Should we do a similar test with splitDepth?

stevennevins replied:
added! [670516b](https://github.com/ethereum-optimism/optimism/pull/17329/commits/670516b6766107c54fde3a206f810a824d817df9)

mbaxter commented on packages/contracts-bedrock/scripts/deploy/DeployImplementations.s.sol:
```diff
@@ -104,6 +115,53 @@ contract DeployImplementations is Script {
+            // Validate V2 game depth parameters are sensible
+            require(
+                _input.faultGameV2MaxGameDepth > 0 && _input.faultGameV2MaxGameDepth <= 200,
+                "DeployImplementations: faultGameV2MaxGameDepth out of valid range (1-200)"
```
I think the max is 125?:
```suggestion
                _input.faultGameV2MaxGameDepth > 0 && _input.faultGameV2MaxGameDepth <= 125,
                "DeployImplementations: faultGameV2MaxGameDepth out of valid range (1-125)"
```

stevennevins replied:
Updated [3c39c2b](https://github.com/ethereum-optimism/optimism/pull/17329/commits/3c39c2b887c52712511b1c0e3d2a30d484b61af0)

mbaxter commented on packages/contracts-bedrock/scripts/deploy/DeployImplementations.s.sol:
```diff
@@ -104,6 +115,53 @@ contract DeployImplementations is Script {
+            deployPermissionedDisputeGameV2Impl(_input, output_);
+        } else {
+            // When V2 is not enabled, ensure V2 parameters are either zero or within safe bounds
+            // to prevent issues if the flag is later enabled
```
(nit) Not sure we need to validate when v2 is disabled

stevennevins replied:
agree, removed [150ccb5](https://github.com/ethereum-optimism/optimism/pull/17329/commits/150ccb5c727573b2c84365faebfa45fe3052450d)

mbaxter commented on packages/contracts-bedrock/test/opcm/DeployImplementations.t.sol:
```diff
@@ -78,36 +86,50 @@ contract DeployImplementations_Test is Test {
-            })
         );
 
-        ISuperchainConfig superchainConfigImpl = ISuperchainConfig(_superchainConfigImpl);
-        vm.prank(address(superchainProxyAdmin));
-        IProxy(payable(address(superchainConfigProxy))).upgradeTo(address(superchainConfigImpl));
```
Why did we cut this superchainConfig deployment logic?  I'm wondering if the intention was to set the implementation to some empty contract, then later verify that we set it to an actual contract?

stevennevins replied:
I think I thought it wasn't being used but i see now it's passed as a parameter and comes from contract state

stevennevins replied:
I think it's a bit strange to fuzz the upgrade with a contract with an address that has no code

stevennevins replied:
Added back [31a635b](https://github.com/ethereum-optimism/optimism/pull/17329/commits/31a635be5a1abb1e91147507cb12b26c931660bf)

mbaxter reviewed (commented):
Looks good - left a few comments, mostly nits.  And just calling out that we talked about guarding blueprint deployment for the dispute games based on the v2 flag, but we can do that in a follow-up PR.

stevennevins committed:
test: remove constraints on bitmap

stevennevins committed:
fix: remove input validation from deploy implemenations

stevennevins committed:
Revert "fix: remove input validation from deploy implemenations"

This reverts commit aa2557bf164854b9b917601fd32b8b542d3eef53.

stevennevins committed:
test: add splitdepth check

stevennevins committed:
test: update run with feature flag on and update asserts

stevennevins committed:
fix: add back superchain proxy admin and superchainConfigProxy

stevennevins committed:
fix: correct max game depth

stevennevins committed:
test: expand bounded values

stevennevins committed:
refactor: move asserts about v2 config params

mbaxter commented on op-chain-ops/interopgen/deploy.go:
```diff
@@ -196,6 +196,10 @@ func DeploySuperchainToL1(l1Host *script.Host, opcmScripts *opcm.Scripts, superC
+		FaultGameV2MaxGameDepth:         big.NewInt(73),
+		FaultGameV2SplitDepth:           big.NewInt(30),
+		FaultGameV2ClockExtension:       big.NewInt(10800),
+		FaultGameV2MaxClockDuration:     big.NewInt(302400),
```
Just noticed these parameters should be configurable - opened a follow-up ticket [here](https://github.com/ethereum-optimism/optimism/issues/17614).
